integration/client/container_idmapped_linux_test.go

```diff
@@ -23,101 +23,155 @@ import (
+		idMap   userns.IDMap
+		snapOpt func(idMap userns.IDMap) snapshots.Opt
+		expUID  uint32
+		expGID  uint32
+	}{
+		"TestIDMappedOverlay-SingleMapping": {
+			idMap: userns.IDMap{
+				UidMap: []specs.LinuxIDMapping{
+					{
+						ContainerID: 0,
+						HostID:      33,
+						Size:        65535,
+					},
+				},
+				GidMap: []specs.LinuxIDMapping{
+					{
+						ContainerID: 0,
+						HostID:      33,
+						Size:        65535,
+					},
+				},
+			},
+			snapOpt: func(idMap userns.IDMap) snapshots.Opt {
+				return containerd.WithRemapperLabels(
+					idMap.UidMap[0].ContainerID, idMap.UidMap[0].HostID,
+					idMap.GidMap[0].ContainerID, idMap.GidMap[0].HostID,
+					idMap.UidMap[0].Size)
+			},
+			expUID: 33,
+			expGID: 33,
+		},
+		"TestIDMappedOverlay-MultiMapping": {
+			idMap: userns.IDMap{
+				UidMap: []specs.LinuxIDMapping{
+					{
+						ContainerID: 0,
+						HostID:      33,
+						Size:        100,
+					},
+					{
+						ContainerID: 100,
+						HostID:      333,
+						Size:        65536,
+					},
+				},
+				GidMap: []specs.LinuxIDMapping{
+					{
+						ContainerID: 0,
+						HostID:      66,
+						Size:        100,
+					},
+					{
+						ContainerID: 100,
+						HostID:      666,
+						Size:        65536,
+					},
+				},
+			},
+			snapOpt: func(idMap userns.IDMap) snapshots.Opt {
+				return containerd.WithUserNSRemapperLabels(idMap.UidMap, idMap.GidMap)
+			},
+			expUID: 33,
+			expGID: 66,
+		},
+	} {
+		var (
+			upperPath   string
+			lowerPaths  []string
+			snapshotter = "overlayfs"
+			ctx, cancel = testContext(t)
+			id          = name
+		)
+		defer cancel()
+
+		client, err := newClient(t, address)
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer client.Close()
 
-	container, err := client.NewContainer(ctx, id,
-		containerd.WithImage(image),
-		containerd.WithImageConfigLabels(image),
-		containerd.WithSnapshotter(snapshotter),
-		containerd.WithNewSnapshot(id, image, containerd.WithRemapperLabels(uidMap.ContainerID, uidMap.HostID, gidMap.ContainerID, gidMap.HostID, length)),
-		containerd.WithNewSpec(oci.WithImageConfig(image),
-			oci.WithUserNamespace([]specs.LinuxIDMapping{uidMap}, []specs.LinuxIDMapping{gidMap}),
-			longCommand))
-	if err != nil {
-		t.Fatal(err)
-	}
-	defer container.Delete(ctx, containerd.WithSnapshotCleanup)
+		image, err := client.Pull(ctx, testMultiLayeredImage, containerd.WithPullUnpack)
+		if err != nil {
+			t.Fatal(err)
+		}
+		t.Logf("image %s pulled!", testMultiLayeredImage)
+
+		container, err := client.NewContainer(ctx, id,
+			containerd.WithImage(image),
+			containerd.WithImageConfigLabels(image),
+			containerd.WithSnapshotter(snapshotter),
+			containerd.WithNewSnapshot(id, image, test.snapOpt(test.idMap)),
+			containerd.WithNewSpec(oci.WithImageConfig(image),
+				oci.WithUserNamespace(test.idMap.UidMap, test.idMap.GidMap),
+				longCommand))
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer container.Delete(ctx, containerd.WithSnapshotCleanup)
```

Comment:
defer in loop doesn't seem to work once you use the same name. maybe you should add `t.Run` function for each run.